### PR TITLE
.NET support in R2

### PR DESCRIPTION
**Pull request by lionaneesh**
Work in progress.

### CODE REVIEW

**Comment by Maijin** [comment]
Niiiiiice bro 👍

File: libr/bin/format/pe/pe.c [hunk 1]
```diff
@@ -619,6 +620,140 @@ int PE_(bin_pe_get_actual_checksum)(struct PE_(r_bin_pe_obj_t) *bin) {
...
+		printf("%s %d\n", stream_name, c);
+		stream->Name = stream_name;
+		streams[count] = stream;
+		start_of_stream += 8 + c;
+		count += 1;
+	}
+	bin->streams = streams;
+	return 1;
+fail:
+	eprintf ("Warning: read (metaadata header)\n");
```
Comment (Maijin) [inline libr/bin/format/pe/pe.c]: typo

**Comment by Maijin** [comment]
This is linked to https://github.com/radare/radare2/issues/1191

**Comment by radare** [comment]
Use the new `r2-indent` command to indent the code before sending a PR :)

File: libr/bin/format/pe/pe.c [hunk 2]
```diff
@@ -619,6 +620,140 @@ int PE_(bin_pe_get_actual_checksum)(struct PE_(r_bin_pe_obj_t) *bin) {
...
+
+	while (count < metadata->NumberOfStreams) {
+		stream = malloc(sizeof(PE_(image_metadata_stream)));
+		if (!stream) goto fail;
+
+		if (r_buf_fread_at (bin->b, start_of_stream, (ut8*)stream, bin->big_endian ? "2I": "2i", 1) < 1) {
+			free (stream);
+			goto fail;
+		}
+		printf("DirectoryAddress: %x Size: %x\n", stream->Offset, stream->Size);
```
Comment (radare) [inline libr/bin/format/pe/pe.c]: use eprintf.

File: libr/bin/format/pe/pe.c [hunk 3]
```diff
@@ -619,6 +620,140 @@ int PE_(bin_pe_get_actual_checksum)(struct PE_(r_bin_pe_obj_t) *bin) {
...
+
+	// read metadata streams
+	int start_of_stream = metadata_directory + 20 + metadata->VersionStringLength;
+	PE_(image_metadata_stream) *stream;
+	PE_(image_metadata_stream) **streams = malloc(sizeof(PE_(image_metadata_stream) *) * metadata->NumberOfStreams);
+	if (!streams) goto fail;
+	int count = 0;
+
+	while (count < metadata->NumberOfStreams) {
+		stream = malloc(sizeof(PE_(image_metadata_stream)));
```
Comment (radare) [inline libr/bin/format/pe/pe.c]: use R_NEW0

File: libr/bin/format/pe/pe.c [hunk 4]
```diff
@@ -619,6 +620,140 @@ int PE_(bin_pe_get_actual_checksum)(struct PE_(r_bin_pe_obj_t) *bin) {
...
+		printf(".NET Version: %s\n", metadata->VersionString);
+	}
+
+	// read the header after the string
+	rr = r_buf_fread_at (bin->b, metadata_directory + 16 + metadata->VersionStringLength,
+						(ut8*)(&metadata->Flags), bin->big_endian ? "2S": "2s", 1);
+
+	if (rr < 1) goto fail;
+
+	printf("Number of Metadata Streams: %d\n", metadata->NumberOfStreams);
```
Comment (radare) [inline libr/bin/format/pe/pe.c]: use eprintf

File: libr/bin/format/pe/pe.c [hunk 5]
```diff
@@ -619,6 +620,140 @@ int PE_(bin_pe_get_actual_checksum)(struct PE_(r_bin_pe_obj_t) *bin) {
...
+		rr = r_buf_read_at (bin->b, metadata_directory + 16,
+						    (ut8*)(metadata->VersionString), len);
+		if (rr != len) {
+			eprintf ("Warning: read (metaadata header) - cannot parse version string\n");
+			free (metadata->VersionString);
+			free (metadata);
+			return 0;
+		}
+
+		printf(".NET Version: %s\n", metadata->VersionString);
```
Comment (radare) [inline libr/bin/format/pe/pe.c]: eprintf

File: libr/bin/format/pe/pe.c [hunk 6]
```diff
@@ -619,6 +620,140 @@ int PE_(bin_pe_get_actual_checksum)(struct PE_(r_bin_pe_obj_t) *bin) {
...
+	rr = r_buf_fread_at (bin->b, metadata_directory + 8,
+					    (ut8*)(&metadata->Reserved), bin->big_endian ? "1I": "1i", 1);
+	if (rr < 1) goto fail;
+
+	rr = r_buf_fread_at (bin->b, metadata_directory + 12,
+					    (ut8*)(&metadata->VersionStringLength), bin->big_endian ? "1I": "1i", 1);
+	if (rr < 1) goto fail;
+
+
+	printf("Metadata Signature: %x %x %d\n", metadata_directory, metadata->Signature, metadata->VersionStringLength);
```
Comment (radare) [inline libr/bin/format/pe/pe.c]: more eprintf

**Comment by lionaneesh** [comment]
Seems good now? @radare

File: libr/bin/format/pe/pe_specs.h [hunk 7]
```diff
@@ -234,6 +234,45 @@ typedef struct {
...
+	ut16 MinorRuntimeVersion;
+	ut32 MetaDataDirectoryAddress;
+	ut32 MetaDataDirectorySize;
+	ut32 Flags;
+	ut32 EntryPointToken;
+	ut32 ResourcesDirectoryAddress;
+	ut32 ResourcesDirectorySize;
+	ut32 StrongNameSignatureAddress;
+	ut32 StrongNameSignatureSize;
+    ut32 CodeManagerTableAddress;
```
Comment (radare) [inline libr/bin/format/pe/pe_specs.h]: spaces here. use r2-indent to avoid those issues

File: libr/bin/format/pe/pe.h [hunk 8]
```diff
@@ -88,6 +94,7 @@ struct PE_(r_bin_pe_obj_t) {
 	int num_sections;
 	int endian;
 	bool verbose;
+	int big_endian;
```
Comment (radare) [inline libr/bin/format/pe/pe.h]: use bool instead of int

File: libr/bin/format/pe/pe.c [hunk 9]
```diff
@@ -619,6 +620,140 @@ int PE_(bin_pe_get_actual_checksum)(struct PE_(r_bin_pe_obj_t) *bin) {
...
+	}
+	bin->streams = streams;
+	return 1;
+fail:
+	eprintf ("Warning: read (metadata header)\n");
+	free (metadata);
+	return 0;
+}
+
+static int bin_pe_init_clr_hdr(struct PE_(r_bin_pe_obj_t) *bin) {
```
Comment (radare) [inline libr/bin/format/pe/pe.c]: return bool here

File: libr/bin/format/pe/pe.c [hunk 10]
```diff
@@ -619,6 +620,140 @@ int PE_(bin_pe_get_actual_checksum)(struct PE_(r_bin_pe_obj_t) *bin) {
...
+	PE_(image_data_directory) *clr_dir = &bin->data_directory[PE_IMAGE_DIRECTORY_ENTRY_COM_DESCRIPTOR];
+	PE_DWord image_clr_hdr_paddr       = clr_dir ? bin_pe_rva_to_paddr (bin, clr_dir->VirtualAddress) : 0;
+	int clr_dir_size                   = clr_dir ? clr_dir->Size : 0;
+	PE_(image_clr_header) *clr_hdr 	   = R_NEW0 (PE_(image_clr_header));
+	int rr, len 					   = sizeof (PE_(image_clr_header));
+
+	if (!clr_hdr) return 0;
```
Comment (radare) [inline libr/bin/format/pe/pe.c]: missing braces, use r2-indent to avoid that

File: libr/bin/format/pe/pe.c [hunk 11]
```diff
@@ -619,6 +620,140 @@ int PE_(bin_pe_get_actual_checksum)(struct PE_(r_bin_pe_obj_t) *bin) {
...
+	rr = r_buf_read_at (bin->b, image_clr_hdr_paddr,
```
Comment (radare) [inline libr/bin/format/pe/pe.c]: wrong indent . put everything in 1 line

File: libr/bin/format/pe/pe.c [hunk 12]
```diff
@@ -619,6 +620,140 @@ int PE_(bin_pe_get_actual_checksum)(struct PE_(r_bin_pe_obj_t) *bin) {
...
+	free (metadata);
+	return 0;
+}
+
+static int bin_pe_init_clr_hdr(struct PE_(r_bin_pe_obj_t) *bin) {
+	PE_(image_data_directory) *clr_dir = &bin->data_directory[PE_IMAGE_DIRECTORY_ENTRY_COM_DESCRIPTOR];
+	PE_DWord image_clr_hdr_paddr       = clr_dir ? bin_pe_rva_to_paddr (bin, clr_dir->VirtualAddress) : 0;
+	int clr_dir_size                   = clr_dir ? clr_dir->Size : 0;
+	PE_(image_clr_header) *clr_hdr 	   = R_NEW0 (PE_(image_clr_header));
+	int rr, len 					   = sizeof (PE_(image_clr_header));
```
Comment (radare) [inline libr/bin/format/pe/pe.c]: too many spaces between varname and =

File: libr/bin/format/pe/pe.c [hunk 13]
```diff
@@ -619,6 +620,140 @@ int PE_(bin_pe_get_actual_checksum)(struct PE_(r_bin_pe_obj_t) *bin) {
...
+	return 1;
+fail:
+	eprintf ("Warning: read (metadata header)\n");
+	free (metadata);
+	return 0;
+}
+
+static int bin_pe_init_clr_hdr(struct PE_(r_bin_pe_obj_t) *bin) {
+	PE_(image_data_directory) *clr_dir = &bin->data_directory[PE_IMAGE_DIRECTORY_ENTRY_COM_DESCRIPTOR];
+	PE_DWord image_clr_hdr_paddr       = clr_dir ? bin_pe_rva_to_paddr (bin, clr_dir->VirtualAddress) : 0;
```
Comment (radare) [inline libr/bin/format/pe/pe.c]: dont align at =

File: libr/bin/format/pe/pe.c [hunk 14]
```diff
@@ -619,6 +620,140 @@ int PE_(bin_pe_get_actual_checksum)(struct PE_(r_bin_pe_obj_t) *bin) {
...
+			goto fail;
+		}
+		eprintf("DirectoryAddress: %x Size: %x\n", stream->Offset, stream->Size);
+		char *stream_name = malloc(MAX_METADATA_STRING_LENGTH + 1);
+		int c = bin_pe_read_metadata_string(stream_name, bin->b->buf + start_of_stream + 8);
+		if (c == 0) {
+			free(stream);
+			goto fail;
+		}
+		eprintf("Stream name: %s %d\n", stream_name, c);
```
Comment (radare) [inline libr/bin/format/pe/pe.c]: missing space before (. use r2-indent to avoid that

File: libr/bin/format/pe/pe.c [hunk 15]
```diff
@@ -619,6 +620,140 @@ int PE_(bin_pe_get_actual_checksum)(struct PE_(r_bin_pe_obj_t) *bin) {
...
+
+	while (count < metadata->NumberOfStreams) {
+		stream = R_NEW0(sizeof(PE_(image_metadata_stream)));
+		if (!stream) goto fail;
+
+		if (r_buf_fread_at (bin->b, start_of_stream, (ut8*)stream, bin->big_endian ? "2I": "2i", 1) < 1) {
+			free (stream);
+			goto fail;
+		}
+		eprintf("DirectoryAddress: %x Size: %x\n", stream->Offset, stream->Size);
```
Comment (radare) [inline libr/bin/format/pe/pe.c]: same here

File: libr/bin/format/pe/pe.c [hunk 16]
```diff
@@ -619,6 +620,140 @@ int PE_(bin_pe_get_actual_checksum)(struct PE_(r_bin_pe_obj_t) *bin) {
...
+		char *stream_name = malloc(MAX_METADATA_STRING_LENGTH + 1);
```
Comment (radare) [inline libr/bin/format/pe/pe.c]: check null  and add space before (

File: libr/bin/format/pe/pe.c [hunk 17]
```diff
@@ -619,6 +620,140 @@ int PE_(bin_pe_get_actual_checksum)(struct PE_(r_bin_pe_obj_t) *bin) {
...
+
+	// read metadata streams
+	int start_of_stream = metadata_directory + 20 + metadata->VersionStringLength;
+	PE_(image_metadata_stream) *stream;
+	PE_(image_metadata_stream) **streams = malloc(sizeof(PE_(image_metadata_stream) *) * metadata->NumberOfStreams);
+	if (!streams) goto fail;
+	int count = 0;
+
+	while (count < metadata->NumberOfStreams) {
+		stream = R_NEW0(sizeof(PE_(image_metadata_stream)));
```
Comment (radare) [inline libr/bin/format/pe/pe.c]: spaces before (

File: libr/bin/format/pe/pe.c [hunk 18]
```diff
@@ -619,6 +620,140 @@ int PE_(bin_pe_get_actual_checksum)(struct PE_(r_bin_pe_obj_t) *bin) {
...
+	if (rr < 1) goto fail;
+
+	eprintf("Number of Metadata Streams: %d\n", metadata->NumberOfStreams);
+	bin->metadata_header = metadata;
+
+
+	// read metadata streams
+	int start_of_stream = metadata_directory + 20 + metadata->VersionStringLength;
+	PE_(image_metadata_stream) *stream;
+	PE_(image_metadata_stream) **streams = malloc(sizeof(PE_(image_metadata_stream) *) * metadata->NumberOfStreams);
```
Comment (radare) [inline libr/bin/format/pe/pe.c]: use R_NEW0

File: libr/bin/format/pe/pe.c [hunk 19]
```diff
@@ -619,6 +620,140 @@ int PE_(bin_pe_get_actual_checksum)(struct PE_(r_bin_pe_obj_t) *bin) {
...
+		}
+
+		eprintf(".NET Version: %s\n", metadata->VersionString);
+	}
+
+	// read the header after the string
+	rr = r_buf_fread_at (bin->b, metadata_directory + 16 + metadata->VersionStringLength,
+						(ut8*)(&metadata->Flags), bin->big_endian ? "2S": "2s", 1);
+
+	if (rr < 1) goto fail;
```
Comment (radare) [inline libr/bin/format/pe/pe.c]: braces or r2-indent

File: libr/bin/format/pe/pe.c [hunk 20]
```diff
@@ -619,6 +620,140 @@ int PE_(bin_pe_get_actual_checksum)(struct PE_(r_bin_pe_obj_t) *bin) {
...
+		rr = r_buf_read_at (bin->b, metadata_directory + 16,
+							(ut8*)(metadata->VersionString), len);
+		if (rr != len) {
+			eprintf ("Warning: read (metadata header) - cannot parse version string\n");
+			free (metadata->VersionString);
+			free (metadata);
+			return 0;
+		}
+
+		eprintf(".NET Version: %s\n", metadata->VersionString);
```
Comment (radare) [inline libr/bin/format/pe/pe.c]: missing space

File: libr/bin/format/pe/pe.c [hunk 21]
```diff
@@ -619,6 +620,140 @@ int PE_(bin_pe_get_actual_checksum)(struct PE_(r_bin_pe_obj_t) *bin) {
...
+						(ut8*)metadata, bin->big_endian ? "1I2S": "1i2s", 1);
+	if (rr < 1) goto fail;
+
+	rr = r_buf_fread_at (bin->b, metadata_directory + 8,
+						(ut8*)(&metadata->Reserved), bin->big_endian ? "1I": "1i", 1);
+	if (rr < 1) goto fail;
+
+	rr = r_buf_fread_at (bin->b, metadata_directory + 12,
+						(ut8*)(&metadata->VersionStringLength), bin->big_endian ? "1I": "1i", 1);
+	if (rr < 1) goto fail;
```
Comment (radare) [inline libr/bin/format/pe/pe.c]: braces or r2-indent

File: libr/bin/format/pe/pe.c [hunk 22]
```diff
@@ -619,6 +620,140 @@ int PE_(bin_pe_get_actual_checksum)(struct PE_(r_bin_pe_obj_t) *bin) {
...
+	int covered = 0;
+	while (covered < MAX_METADATA_STRING_LENGTH) {
+		to[covered] = from[covered];
+		if (from[covered] == '\0') { 
+			covered += 1;
+			break;
+		}
+		covered++;
+	}
+	while (covered % 4 != 0) covered++;
```
Comment (radare) [inline libr/bin/format/pe/pe.c]: missing braces

**Comment by lionaneesh** [comment]
ping @radare

**Comment by Maijin** [comment]
eprintf when opening a file ? this is really agressive ❗️ 
image

Can't we have this in `iI` instead ? (related #6785 and sdbization of rbin once again)

**Pull request merged**